### PR TITLE
Fixed DSR loopback connectivity for Windows

### PR DESCRIPTION
**Pull request by zawachte**
**What type of PR is this?**
/kind bugfix

**What this PR does / why we need it**:
Fixed DSR loopback connectivity for Windows

**Does this PR introduce a user-facing change?**:

```release-note
NONE

```

### CODE REVIEW

**Comment by zawachte-msft** [comment]
@madhanrm @ksubrmnn

File: pkg/proxy/winkernel/hnsV1.go [hunk 1]
```diff
@@ -35,6 +35,7 @@ type HostNetworkService interface {
 	deleteEndpoint(hnsID string) error
 	getLoadBalancer(endpoints []endpointsInfo, isILB bool, isDSR bool, sourceVip string, vip string, protocol uint16, internalPort uint16, externalPort uint16) (*loadBalancerInfo, error)
 	deleteLoadBalancer(hnsID string) error
+	updateEndpointPolicy(hnsID string, policy json.RawMessage) error
```
Comment (ksubrmnn) [inline pkg/proxy/winkernel/hnsV1.go]: Can you change hnsID to endpointID so that it is more clear?

File: pkg/proxy/winkernel/hnsV1.go [hunk 2]
```diff
@@ -223,3 +224,7 @@ func (hns hnsV1) deleteLoadBalancer(hnsID string) error {
 	_, err = hnsloadBalancer.Delete()
 	return err
 }
+
+func (hns hnsV1) updateEndpointPolicy(hnsID string, policy json.RawMessage) error {
+	return fmt.Errorf("DSR is not supported in V1. Using non DSR instead")
```
Comment (ksubrmnn) [inline pkg/proxy/winkernel/hnsV1.go]: Change this to updateEndpointPolicy is not implemented for V1. This function should be generic and not specific to DSR

File: pkg/proxy/winkernel/hnsV2.go [hunk 3]
```diff
@@ -89,18 +91,31 @@ func (hns hnsV2) getEndpointByIpAddress(ip string, networkName string) (*endpoin
...
 				hnsID:      endpoint.Id,
 				hns:        hns,
+				policies:   pol,
 			}, nil
 		}
 	}
 
 	return nil, fmt.Errorf("Endpoint %v not found on network %s", ip, networkName)
 }
+func transformToPoliciesinfo(hcnendpointpolicies []hcn.EndpointPolicy) []*policiesinfo {
```
Comment (ksubrmnn) [inline pkg/proxy/winkernel/hnsV2.go]: Rename to policiesToPolicyInfo

File: pkg/proxy/winkernel/hnsV2.go [hunk 4]
```diff
@@ -89,18 +91,31 @@ func (hns hnsV2) getEndpointByIpAddress(ip string, networkName string) (*endpoin
...
+	var endpointPolicies []*policiesinfo
+	for _, po := range hcnendpointpolicies {
+		var policy policiesinfo
```
Comment (ksubrmnn) [inline pkg/proxy/winkernel/hnsV2.go]: Include a log statement here for po

File: pkg/proxy/winkernel/proxier.go [hunk 5]
```diff
@@ -218,7 +237,7 @@ func newServiceInfo(svcPortName proxy.ServicePortName, port *v1.ServicePort, ser
 		stickyMaxAgeSeconds:      stickyMaxAgeSeconds,
 		loadBalancerSourceRanges: make([]string, len(service.Spec.LoadBalancerSourceRanges)),
 		onlyNodeLocalEndpoints:   onlyNodeLocalEndpoints,
-		hns:                      hns,
+		hns: hns,
```
Comment (ksubrmnn) [inline pkg/proxy/winkernel/proxier.go]: Does this pass gofmt?

File: pkg/proxy/winkernel/proxier.go [hunk 6]
```diff
@@ -1025,6 +1044,51 @@ func (proxier *Proxier) syncProxyRules() {
 				newHnsEndpoint, err = hns.getEndpointByIpAddress(ep.ip, hnsNetworkName)
 			}
 
+			if newHnsEndpoint != nil {
+				if newHnsEndpoint.isLocal && proxier.isDSR {
+					var alreadyHasDSRPolicy = false
```
Comment (ksubrmnn) [inline pkg/proxy/winkernel/proxier.go]: Rename to addDSRPolicy

File: pkg/proxy/winkernel/proxier.go [hunk 7]
```diff
@@ -1025,6 +1044,51 @@ func (proxier *Proxier) syncProxyRules() {
...
+					for _, po := range newHnsEndpoint.policies {
+						if strings.Compare(po.Type, "OutBoundNAT") == 0 {
+							var outputSettings outboundnatpolicysetting
+							if err := json.Unmarshal([]byte(po.Settings), &outputSettings); err != nil {
+								break
```
Comment (ksubrmnn) [inline pkg/proxy/winkernel/proxier.go]: Won't this break out of the loop meaning you might not check the other policies?

File: pkg/proxy/winkernel/proxier.go [hunk 8]
```diff
@@ -24,6 +24,7 @@ import (
 	"net"
 	"os"
 	"reflect"
+	"strings"
```
Comment (ksubrmnn) [inline pkg/proxy/winkernel/proxier.go]: You can just use == to compare strings in Go. Is there a reason you want to use strings?

File: pkg/proxy/winkernel/proxier.go [hunk 9]
```diff
@@ -1025,6 +1044,51 @@ func (proxier *Proxier) syncProxyRules() {
 				newHnsEndpoint, err = hns.getEndpointByIpAddress(ep.ip, hnsNetworkName)
 			}
 
+			if newHnsEndpoint != nil {
+				if newHnsEndpoint.isLocal && proxier.isDSR {
```
Comment (ksubrmnn) [inline pkg/proxy/winkernel/proxier.go]: Include a comment for this block explaining that this fixes the DSR loopback issue.

File: pkg/proxy/winkernel/proxier.go [hunk 10]
```diff
@@ -1025,6 +1044,51 @@ func (proxier *Proxier) syncProxyRules() {
...
+							Type:     "OutBoundNAT",
+							Settings: rawJSON,
+						}
+						endpointRequest := policyendpointrequest{
+							Policies: []policiesinfo{newLoopbackPolicy},
+						}
+
+						settingsJson, err := json.Marshal(endpointRequest)
+						if err != nil {
+							break
```
Comment (ksubrmnn) [inline pkg/proxy/winkernel/proxier.go]: Log this error

Comment (ksubrmnn) [inline pkg/proxy/winkernel/proxier.go]: I think this break will skip the logging, ep Info updating in 1150 etc.

**Review by ksubrmnn — CHANGES_REQUESTED** [review]

File: pkg/proxy/winkernel/hnsV2.go (hunk 4 above)
Comment (zawachte-msft) [reply pkg/proxy/winkernel/hnsV2.go]: There is one for policy on line 115 ... but not sure if you were suggesting that we should have a log for both structs?

File: pkg/proxy/winkernel/proxier.go (hunk 8 above)
Comment (zawachte-msft) [reply pkg/proxy/winkernel/proxier.go]: Using == makes me queasy. Especially if this is code that may be running cross-platform. But if that is the go standard... I can change it back.

File: pkg/proxy/winkernel/proxier.go (hunk 5 above)
Comment (zawachte-msft) [reply pkg/proxy/winkernel/proxier.go]: gofmt made this change.

**Comment by feiskyer** [comment]
@zawachte-msft This is cool. I think the changes don't cover https://github.com/kubernetes/kubernetes/pull/62093 yet, right? Could you like to cherry pick the commits together, or let me rebase #62093 after this one?

**Comment by ksubrmnn** [comment]
@feiskyer I think we should keep these PRs separate. Supporting only node local endpoints is different than fixing the loopback issue. 

Also looks like your PR will need rebasing anyway to use the HostNetworkService interface. Can we wait until Zach’s PR goes in and the you can rebase?

**Comment by ksubrmnn** [comment]
@zawachte-msft 

Can you change the description for this PR and the release note?  

WinDSR as a feature flag is already functional and in alpha as of 1.14. The issue is that requests load balanced to the source container (loopback) fail. 

“Implementing DSR” is already handled by using proxies.isDSR when creating loadbalancers. This PR is more of a bug fix.

File: pkg/proxy/winkernel/hnsV2.go (hunk 4 above)
Comment (ksubrmnn) [reply pkg/proxy/winkernel/hnsV2.go]: Yes we should log po (each policy we inspect) as well as the policyInfo we create

File: pkg/proxy/winkernel/proxier.go (hunk 8 above)
Comment (ksubrmnn) [reply pkg/proxy/winkernel/proxier.go]: == is fine. There’s no sort of user input here

**Comment by feiskyer** [comment]
> Also looks like your PR will need rebasing anyway to use the HostNetworkService interface. Can we wait until Zach’s PR goes in and the you can rebase?

sure.

**Pull request closed without merging**